### PR TITLE
Moving code from bin/util into tikv, and crate docs

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

This moves most of the `src/bin/util` code and most of `src/bin/tikv-server.rs` into the `tikv` library. In addition to making it easier to use the library to create a tikv server, it also reduces the amount of utility coded that is recompiled between the three binaries.

This is a follow up to my [previous cleanup pr](https://github.com/tikv/tikv/pull/4513). It does a number of things.

It started out as a documentation patch, and I add some basic crate docs. Then I started documenting some of the configuration methods, but found I wanted more of the setup code from tikv-server, so that it would be easier to write examples.

So then I pulled in some commits for https://github.com/tikv/tikv/issues/4389, moving all of `src/bin/util` into the `tikv` library, and then the setup code from tikv-server. Now it should be easy to show documentation examples for the API.

Moving code from the bins into tikv will reduce the amount of compilation work, but won't reduce serial compilation time.

It is getting pretty big, so I hope to land it before continuing.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

cargo check

## Does this PR affect documentation (docs) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)

Fixes https://github.com/tikv/tikv/issues/4389

### CODE REVIEW

**Comment by siddontang** [comment]
> but won't reduce serial compilation time

Sadly, can it reduce the partial build time?

**Comment by brson** [comment]
> Sadly, can it reduce the partial build time?

No it won't. It's just moving compilation from one module to another. But some code that had to be compiled three times now may need to be compiled once. It probably won't be noticeable.

File: README.md [hunk 1]
```diff
@@ -5,9 +5,9 @@
 ![GitHub release](https://img.shields.io/github/release/tikv/tikv.svg)
 [![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/2574/badge)](https://bestpractices.coreinfrastructure.org/projects/2574)
 
-TiKV ("Ti" stands for Titanium) is an open source distributed transactional key-value database. Unlike other traditional NoSQL systems, TiKV not only provides classical Key-Value APIs, but also transactional APIs with ACID compliance. Built in Rust and powered by Raft, TiKV was originally created to complement [TiDB](https://github.com/pingcap/tidb), a distributed HTAP database compatible with the MySQL protocol.
+TiKV ("Ti" stands for titanium) is an open source distributed transactional key-value database. Unlike other traditional NoSQL systems, TiKV not only provides classical key-value APIs, but also transactional APIs with ACID compliance. Built in Rust and powered by Raft, TiKV was originally created to complement [TiDB](https://github.com/pingcap/tidb), a distributed HTAP database compatible with the MySQL protocol.
```
Comment (siddontang) [inline README.md]: PTAL @lilin90

Comment (lilin90) [reply README.md]: For titanium, I prefer "Titanium". For key-value, I agree with you and we can change it to the lower case version "key-value". @dcalvin What do you think?

Comment (dcalvin) [reply README.md]: I agree with @lilin90. Let's stick with Titanium to be consistent with TiDB usage. And key-value in a lower case makes sense.

Comment (siddontang) [reply README.md]: Friendly ping @brson

**Comment by brson** [comment]
OK, re-merged. Docs say "Titanium" and "key-value" in both places.

**Pull request merged**